magento/magento2
# Fix quote remove deleted products bugs

### Description (*)
In case there are non-existent products in the quote, an exception is thrown in the checkout cart. This pull request fixes that issue by marking non-existent products as "deleted" and hiding them from the front end. "Deleted" products are being removed from the quote when quote "save" is executed.
Additionally, \Magento\Quote\Model\ResourceModel\Quote\Item\Collection::removeItemsWithAbsentProducts method is removed as it's not doing what is expected. Instead of removing non-existent products from the quote, it filters out these products from the collection. That leads to such products never removed from the quote and incorrect quote items count calculations.

### Manual testing scenarios (*)


1. A child product of a configurable product is deleted.

a. Testing steps.

- Add some configurable product configuration to the shopping cart from the front end.
- Delete a **simple** product (added to the cart on the previous step) from the shopping cart using direct SQL query `delete from catalog_product_entity where entity_id = your_entity_id;`. Using a direct SQL query is important because if you delete a product from Magento admin it will trigger the deletion of that product from the quote.
- Go to the "checkout/cart/" page.

b. Actual result

- An exception will be thrown: "Exception #0 (Magento\Framework\Exception\NoSuchEntityException): The product that was requested doesn't exist. Verify the product and try again.".

c. Expected result:

- An error message saying "The requested qty is not available" and "Some item options or their combination are not currently available." must be displayed.
- A customer can not proceed to checkout unless he removes failing product from the cart
- A customer can delete the product which throws errors from the cart
- A customer can delete the product which throws errors from the mini cart

![non_existent_simple](https://user-images.githubusercontent.com/12259690/86370458-730c9500-bc88-11ea-8cbb-a4c8c6ce0383.png)

2. A configurable product is deleted.

a. Testing steps

- Add some configurable product configuration to the shopping cart from the front end.
- Delete a **configurable** product (added to the cart on the previous step) from the shopping cart using direct SQL query `delete from catalog_product_entity where entity_id = your_entity_id;`. Using a direct SQL query is important because if you delete a product from Magento admin it will trigger the deletion of that product from the quote.
- Go to the "checkout/cart/" page.

b. Actual result

- An exception will be thrown: "Exception #0 (Magento\Framework\Exception\NoSuchEntityException): The product that was requested doesn't exist. Verify the product and try again.".

b. Expected result:

- The cart should be empty. 
![empty_cart](https://user-images.githubusercontent.com/12259690/86386510-84f73380-bc9a-11ea-8e43-47b81c6e5d31.png)


### Contribution checklist (*)
